### PR TITLE
Save length check

### LINKED ISSUE #2969: Add max length to the savename
Linux and windows has a max filename of 255, which means 244 characters are usable (without the `.thrivesave` ending). The user can enter longer filenames and the saving crashes.

### PR DESCRIPTION
**Pull request by 84634E1A607A**
**Brief Description of What This PR Does**
Added a save length check.

**Related Issues**
Fixes #2969 
Closes #3028 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: locale/en.po [hunk 1]
```diff
@@ -3773,34 +3773,43 @@ msgstr "Save failed"
...
-msgid "CHOSEN_FILENAME_ALREADY_EXISTS"
+#: ../src/saving/NewSaveMenu.cs:78
+msgid "THE_CHOSEN_FILENAME_ALREADY_EXISTS"
 msgstr "The chosen filename ({0}) already exists. Overwrite?"
 
-#: ../src/saving/NewSaveMenu.cs:172
+#: ../src/saving/NewSaveMenu.cs:112
+msgid "ATTEMPT_TO_WRITE_FILE_FAILED"
+msgstr ""
+"The attempt to write this save file has failed. The save name may be too long. And maybe you don't have write access to the save folder / file."
```
Comment (Kasterisk) [inline locale/en.po:3783]: Line is too long, there should be line breaks created automatically here if you use Poedit. Also probably could be worded like:
```suggestion
"Attempt to write this save file has failed. The save name may be too long or you may not have write access to the save folder/file."
```

File: src/saving/NewSaveMenu.cs [hunk 2]
```diff
@@ -71,7 +76,7 @@ private void ShowOverwriteConfirm(string name)
     {
         // The chosen filename ({0}) already exists. Overwrite?
         overwriteConfirm.DialogText = string.Format(CultureInfo.CurrentCulture,
-            TranslationServer.Translate("CHOSEN_FILENAME_ALREADY_EXISTS"),
+            TranslationServer.Translate("THE_CHOSEN_FILENAME_ALREADY_EXISTS"),
```
Comment (Kasterisk) [inline src/saving/NewSaveMenu.cs:79]: Seems like an unnecessary change.

File: src/saving/FileHelpers.cs [hunk 3]
```diff
@@ -43,4 +44,19 @@ public static bool Exists(string path)
         using var directory = new Directory();
         return directory.FileExists(path);
     }
+
+    public static Error TryCreateWrite(string name)
```
Comment (Kasterisk) [inline src/saving/FileHelpers.cs:48]: Document what's this for.

File: src/saving/NewSaveMenu.cs [hunk 4]
```diff
@@ -101,9 +106,16 @@ private void SaveButtonPressed()
 
     private void OnConfirmSaveName()
     {
-        GUICommon.Instance.PlayButtonPressSound();
+        // Verify name writable
```
Comment (Kasterisk) [inline src/saving/NewSaveMenu.cs:109]:
```suggestion
        // Verify name is writable
```

File: src/saving/NewSaveMenu.cs [hunk 5]
```diff
@@ -101,9 +106,16 @@ private void SaveButtonPressed()
...
+        var name = GetSaveName();
+        if (FileHelpers.TryCreateWrite(name) != Error.Ok)
+        {
+            attemptWriteFailAccept.DialogText = TranslationServer.Translate("ATTEMPT_TO_WRITE_FILE_FAILED");
```
Comment (Kasterisk) [inline src/saving/NewSaveMenu.cs:113]: Why not predefine the dialog text in the scene rather than programmatically?

**Review by Kasterisk — CHANGES_REQUESTED** [review]
The changes to RubySetupSystem may not be good, what is up with that.

Comments:

File: locale/en.po (hunk 1 above)
Comment (hhyyrylainen) [reply locale/en.po]: Actually I just have switched back to not using line wrapping. This time it seems that weblate agrees with that. So finally the translations update PRs won't have a bunch of back and forth with the line wrapping.

File: src/saving/NewSaveMenu.cs (hunk 2 above)
Comment (84634E1A607A) [reply src/saving/NewSaveMenu.cs:79]: I don't know what happened to the exactly. Merge commit did this. Maybe accidentally

File: src/saving/FileHelpers.cs [hunk 6]
```diff
@@ -43,4 +44,19 @@ public static bool Exists(string path)
...
+    {
+        using var file = new File();
+        var e = file.Open(Constants.SAVE_FOLDER + "/" + name, File.ModeFlags.Write);
```
Comment (FuJa0815) [inline src/saving/FileHelpers.cs:51]: Use `Path.Combine`

**Review by FuJa0815 — CHANGES_REQUESTED** [review]

File: src/saving/FileHelpers.cs (hunk 6 above)
Comment (84634E1A607A) [reply src/saving/FileHelpers.cs:51]: What? No such func.

Comment (hhyyrylainen) [reply src/saving/FileHelpers.cs]: It's used elsewhere for example: https://github.com/Revolutionary-Games/Thrive/blob/b11b68630cfdc84c1bd8bd812682bb47f0947e09/src/modding/ModLoader.cs#L65
I'm pretty sure nowhere else is this kind of not fully correct path combining done, elsewhere Path.Combine is used.

Comment (hhyyrylainen) [reply src/saving/FileHelpers.cs]: Turns out we had the old PathUtils from when I didn't know that Path.Combine existed... I made a PR to remove that: https://github.com/Revolutionary-Games/Thrive/pull/3091

File: src/saving/FileHelpers.cs [hunk 7]
```diff
@@ -43,4 +44,22 @@ public static bool Exists(string path)
         using var directory = new Directory();
         return directory.FileExists(path);
     }
+
+    /// <summary>
+    ///   Try to create a file with write access
+    /// </summary>
+    /// <param name="path">Path to file</param>
+    /// <returns>File status</returns>
```
Comment (Kasterisk) [inline src/saving/FileHelpers.cs:52]: I think a better description would be like:
```suggestion
    /// <summary>
    ///   Tests if it is possible to create/write a file on the given path
    /// </summary>
    /// <param name="path">Path to file</param>
    /// <returns>File write access status, <see cref="Error.Ok"/> if successful</returns>
```

**Review by Kasterisk — CHANGES_REQUESTED** [review]

**Review by Kasterisk — APPROVED** [review]
Looks good!

File: src/saving/FileHelpers.cs [hunk 8]
```diff
@@ -43,4 +44,22 @@ public static bool Exists(string path)
...
+
+    /// <summary>
+    ///   Tests if it is possible to create/write a file on the given path
+    /// </summary>
+    /// <param name="path">Path to file</param>
+    /// <returns>File write access status, <see cref="Error.Ok"/> if successful</returns>
+    public static Error TryCreateWrite(string path)
+    {
+        using var file = new File();
+        var e = file.Open(path, File.ModeFlags.Write);
```
Comment (hhyyrylainen) [inline src/saving/FileHelpers.cs]: e should be named something else. This is not one of the contexts where single letter variables are allowed.

File: src/saving/FileHelpers.cs [hunk 9]
```diff
@@ -43,4 +44,22 @@ public static bool Exists(string path)
         using var directory = new Directory();
         return directory.FileExists(path);
     }
+
+    /// <summary>
+    ///   Tests if it is possible to create/write a file on the given path
```
Comment (hhyyrylainen) [inline src/saving/FileHelpers.cs]:
```suggestion
    ///   Tests if it is possible to create/write a file at the given path
```

File: src/saving/NewSaveMenu.cs [hunk 10]
```diff
@@ -102,9 +107,16 @@ private void SaveButtonPressed()
 
     private void OnConfirmSaveName()
     {
-        GUICommon.Instance.PlayButtonPressSound();
```
Comment (hhyyrylainen) [inline src/saving/NewSaveMenu.cs]: Shouldn't the button press sound still be played?

Comment (84634E1A607A) [reply src/saving/NewSaveMenu.cs]: This function is only called from one function above,  in which the button press sound is already played.

Comment (hhyyrylainen) [reply src/saving/NewSaveMenu.cs]: That's fine, then.

File: src/saving/FileHelpers.cs [hunk 11]
```diff
@@ -43,4 +44,23 @@ public static bool Exists(string path)
         using var directory = new Directory();
         return directory.FileExists(path);
     }
+
+    /// <summary>
+    ///   Tests if it is possible to create/write a file at the given path
+    /// </summary>
+    /// <param name="path">Path to file</param>
+    /// <returns>File write access status, <see cref="Error.Ok"/> if successful</returns>
+    /// <remarks>THIS FUNCTION IS DESTRUCTIVE! MAKE SURE THE FILE TO TEST DOESN'T EXIST!</remarks>
```
Comment (hhyyrylainen) [inline src/saving/FileHelpers.cs:53]: Check the right format in style guide for remarks comments.

**Comment by hhyyrylainen** [comment]
I saw one style issue. I guess we have tested the functionality enough already? So once that one slight style issue and merge conflicts are solved, I'll just merge this to get this finished.

File: locale/en.po [hunk 12]
```diff
@@ -3820,34 +3820,42 @@ msgstr "Save failed"
...
 msgid "OVERWRITE_EXISTING_SAVE"
 msgstr "Overwrite existing save:"
 
-#: ../src/saving/NewSaveMenu.tscn:116
+#: ../src/saving/NewSaveMenu.tscn:117
 msgid "OVERWRITE_EXISTING_SAVE_PROMPT"
 msgstr "Overwrite existing save?"
 
+#: ../src/saving/NewSaveMenu.tscn:121
+msgid "CANNOT_WRITE"
```
Comment (hhyyrylainen) [inline locale/en.po:3852]: These translation keys should have "SAVE" in them as these are very explicitly meant for save errors not generic can't write a file error. So rename this "CANNOT_WRITE_SAVE" and the other "ATTEMPT_TO_WRITE_SAVE_FAILED"

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

**Comment by hhyyrylainen** [comment]
You only changed one of the two translation keys.

**Comment by 84634E1A607A** [comment]
Well, the last CI error is rather strange.

`Critical error in processing websocket connection, no further output will be received`

It isn't code-related.

**Comment by hhyyrylainen** [comment]
It isn't and you should just hit the button to retry failed jobs.

Also if anyone at all wants to help with the devcenter bugs, that's very much welcome.

**Comment by 84634E1A607A** [comment]
Btw the text should actually be like this. So I changed it.

**Comment by 84634E1A607A** [comment]
> just hit the button to retry failed jobs

Well, this is the first time that I realized there's a button to rerun the checks.

File: locale/en.po [hunk 13]
```diff
@@ -3820,34 +3820,42 @@ msgstr "Save failed"
...
 msgid "OVERWRITE_EXISTING_SAVE_PROMPT"
 msgstr "Overwrite existing save?"
 
+#: ../src/saving/NewSaveMenu.tscn:121
+msgid "CANNOT_WRITE_SAVE"
+msgstr "Cannot Write Save"
+
+#: ../src/saving/NewSaveMenu.tscn:125
+msgid "ATTEMPT_TO_WRITE_FILE_FAILED"
+msgstr "Attempt to write this save file has failed. The save name may be too long or you may not have write access to the save folder."
```
Comment (hhyyrylainen) [inline locale/en.po:3857]: It still says "save" here and the key is not changed to have "save" in it.

**Review by hhyyrylainen — APPROVED** [review]
